### DISCUSSION TITLE
Implementing initial condition as hard constraint using values obtained from interpolation between arbitrary values

### DISCUSSION THREAD

**Discussion by Rahul-Patel25** [lululxvi/deepxde#1265] [Q&A]
Hello,
Thanks for this useful library.
 
I am working on the [heat equation](https://deepxde.readthedocs.io/en/latest/demos/pinn_forward/heat.html) example. 

To incorporate the initial condition as a soft constraint I can use 

```
ic = dde.icbc.IC(
    geomtime,
    lambda x: np.sin(np.pi * x[:, 0:1]),
    lambda _, on_initial: on_initial,
)
```

or to implement the same initial condition as hard constraint I can transform the output as 
```
def output_transform(x, y):
    return x[:, 0:1] * (1 - x[:, 0:1]) * y * x[:, 1:2] +  tf.sin(pi * x[:, 0])
```
and will get the same results. 

Now, rather than specifying the sinusoid function in the ICs, I want to specify arbitrary points (let’s say 7) of X and specify the corresponding solution (y) at that point for the initial state (t=0). For intermediate values of x, the values of y should be obtained via interpolation (the end goal is to pass these arbitrary solution values (y) at t=0 as an input to the Neural network to implement arbitrary initial condition).  

For soft constraints, I was able to implement them as follows.
```
def initial_funcval(x):
    size_vector = x[:, 0:1].shape[0]
    outval = np.zeros([size_vector])
    xx = np.linspace(0, 1, 7)
    yy = [0, 0.5,0.866,1, 0.866,0.5,0]

    for i in range(size_batch):
                f = interpolate.interp1d(xx, yy, kind='linear')
        outval[i] = f(x[i, 0:1])

    return outval.reshape(size_vector, 1)
```

```
ic = dde.icbc.IC(
    geomtime,
    lambda x: initial_funcval(x),
    lambda _, on_initial: on_initial,
)
```
`interpolate.interp1d` is an 1D interpolation function of SciPy. 

How can I implement the same as hard constraints by transforming the output layer? Could you please guide me through it.

**Comment by lululxvi** [comment]
You cannot use SciPy. You need to implement it in TensoFlow/pyTorch/etc.
